rules/recording.go

```diff
@@ -73,8 +76,60 @@ func (rule *RecordingRule) Labels() labels.Labels {
 
 // Eval evaluates the rule and then overrides the metric names and labels accordingly.
 func (rule *RecordingRule) Eval(ctx context.Context, ts time.Time, query QueryFunc, _ *url.URL, limit int) (promql.Vector, error) {
-	ctx = NewOriginContext(ctx, NewRuleDetail(rule))
+	vector, err := rule.eval(ctx, ts, query, nil, limit)
+	if err != nil {
+		return nil, err
+	}
+
+	rule.SetHealth(HealthGood)
+	rule.SetLastError(err)
+	return vector, nil
+}
+
+// EvalWithExemplars evaluates the rule and then overrides the metric names and labels accordingly along with emitting exemplars.
+func (rule *RecordingRule) EvalWithExemplars(ctx context.Context, ts time.Time, interval time.Duration,
+	query QueryFunc, eq storage.ExemplarQuerier, _ *url.URL, limit int) (promql.Vector, []exemplar.QueryResult, error) {
+	vector, err := rule.eval(ctx, ts, query, nil, limit)
+	if err != nil {
+		return nil, nil, err
+	}
 
+	selectors := parser.ExtractSelectors(rule.vector)
+	if len(selectors) < 1 {
+		return nil, nil, err
+	}
+
+	// Query all the raw exemplars that match the query
+	ex, err := eq.Select(timestamp.FromTime(ts.Add(-1*interval)), timestamp.FromTime(ts), selectors...)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// For each vector received, we create matchers for all labels of the vector minus the metric name. If it matches
+	// we replace the exemplar's series labels with the vector's labels. We do this to ensure that we are able to append
+	// the exemplars to the same series ref the recorded metric went into.
+	for _, v := range vector {
+		matchers := make([]*labels.Matcher, 0, len(v.Metric))
+		for _, l := range v.Metric {
+			if l.Name != labels.MetricName {
+				matchers = append(matchers, labels.MustNewMatcher(labels.MatchEqual, l.Name, l.Value))
+			}
+		}
+
+		for i := range ex {
+			if ok := matches(ex[i].SeriesLabels, matchers); ok {
+				ex[i].SeriesLabels = v.Metric
+			}
```

Comment:
What happens for exemplars that never match?

Reply:
modified the code so that we drop them on the floor.